### PR TITLE
Fix COPY FROM resulting in dulicate data due to consecutive SocketTimeout

### PR DESCRIPTION
**Pull request by jeeminso**
## Summary of the changes / Why this improves CrateDB

There was a bug in `FileReadingIterator.getLine(...)`.

EX) there was a `SocketTimeoutException` at 100th line. Then there will be a recursive call to `getLine(reader, 100 + 1, 1)` -- (1). While executing this call, another `SocketTimeoutException` happened at 10th line, causing another recursive call to `getLine(reader, 10 + 1, 1)` -- (2). Here is the problem, (1) will return whatever is returned by (2) which is the data at 10th line.  

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: server/src/test/java/io/crate/execution/engine/collect/files/FileReadingIteratorTest.java [hunk 1]
```diff
@@ -232,6 +234,85 @@ public String readLine() throws IOException {
         tester.verifyResultAndEdgeCaseBehaviour(expectedResult);
     }
 
+    @Test
+    public void test_socket_timeouts_handled_by_retries() throws Exception {
```
Comment (seut) [inline server/src/test/java/io/crate/execution/engine/collect/files/FileReadingIteratorTest.java:238]:
```suggestion
    public void test_consecutive_retries_will_not_result_in_duplicate_reads() throws Exception {
```

File: server/src/test/java/io/crate/execution/engine/collect/files/FileReadingIteratorTest.java [hunk 2]
```diff
@@ -232,6 +234,85 @@ public String readLine() throws IOException {
...
+        ArrayList<String> fileUris = new ArrayList<>();
+
+        var tempFilePath1 = createTempFile("tempfile1", ".csv");
+        try (OutputStreamWriter writer = new OutputStreamWriter(new FileOutputStream(tempFilePath1.toFile()),
+                                                                StandardCharsets.UTF_8)) {
+            writer.write("name,id,age\n");
+            writer.write("Arthur1,4,38\nArthur2,4,38\nArthur3,4,38\nArthur4,4,38\nArthur5,4,38\n");
```
Comment (seut) [inline server/src/test/java/io/crate/execution/engine/collect/files/FileReadingIteratorTest.java:245]:
```suggestion
            writer.write("id\n");
            writer.write("1\n2\n3\n4\n5\n");
```

File: server/src/test/java/io/crate/execution/engine/collect/files/FileReadingIteratorTest.java [hunk 3]
```diff
@@ -232,6 +234,85 @@ public String readLine() throws IOException {
...
+        }
+        fileUris.add(tempFilePath1.toUri().toString());
+
+        var tempFilePath2 = createTempFile("tempfile2", ".csv");
+        try (OutputStreamWriter writer = new OutputStreamWriter(new FileOutputStream(tempFilePath2.toFile()),
+                                                                StandardCharsets.UTF_8)) {
+            writer.write("name,id,age\n");
+            writer.write("Trillian1,5,33\nTrillian2,5,33\nTrillian3,5,33\nTrillian4,5,33\nTrillian5,5,33\n");
+        }
+        fileUris.add(tempFilePath2.toUri().toString());
```
Comment (seut) [inline server/src/test/java/io/crate/execution/engine/collect/files/FileReadingIteratorTest.java:255]: Lets remove this 2nd file as it is not needed to validate the bug.

File: server/src/test/java/io/crate/execution/engine/collect/files/FileReadingIteratorTest.java [hunk 4]
```diff
@@ -232,6 +234,85 @@ public String readLine() throws IOException {
...
+            () -> new FileReadingIterator(
+                fileUris,
+                inputs,
+                ctx.expressions(),
+                null,
+                Map.of(LocalFsFileInputFactory.NAME, new LocalFsFileInputFactory()),
+                false,
+                1,
+                0,
+                List.of("name", "id", "age"),
```
Comment (seut) [inline server/src/test/java/io/crate/execution/engine/collect/files/FileReadingIteratorTest.java:273]:
```suggestion
                List.of("id"),
```

File: server/src/test/java/io/crate/execution/engine/collect/files/FileReadingIteratorTest.java [hunk 5]
```diff
@@ -232,6 +234,85 @@ public String readLine() throws IOException {
...
+
+                        private int currentLineNumber = 0;
+
+                        @Override
+                        public String readLine() throws IOException {
+                            var line = super.readLine();
+                            if (new Random().nextBoolean()) {
+                                // current implementation does not handle SocketTimeoutException thrown when parsing header so skip it here as well.
+                                if (currentLineNumber++ > 0 &&
+                                    new Random().nextBoolean() && retry++ < MAX_SOCKET_TIMEOUT_RETRIES) {
```
Comment (seut) [inline server/src/test/java/io/crate/execution/engine/collect/files/FileReadingIteratorTest.java:292]: Lets make the test deterministic
```suggestion
                                    retry++ < MAX_SOCKET_TIMEOUT_RETRIES) {
```

File: server/src/test/java/io/crate/execution/engine/collect/files/FileReadingIteratorTest.java [hunk 6]
```diff
@@ -232,6 +234,85 @@ public String readLine() throws IOException {
...
+            new Object[]{"{\"name\":\"Arthur1\",\"id\":\"4\",\"age\":\"38\"}"},
+            new Object[]{"{\"name\":\"Arthur2\",\"id\":\"4\",\"age\":\"38\"}"},
+            new Object[]{"{\"name\":\"Arthur3\",\"id\":\"4\",\"age\":\"38\"}"},
+            new Object[]{"{\"name\":\"Arthur4\",\"id\":\"4\",\"age\":\"38\"}"},
+            new Object[]{"{\"name\":\"Arthur5\",\"id\":\"4\",\"age\":\"38\"}"},
+            new Object[]{"{\"name\":\"Trillian1\",\"id\":\"5\",\"age\":\"33\"}"},
+            new Object[]{"{\"name\":\"Trillian2\",\"id\":\"5\",\"age\":\"33\"}"},
+            new Object[]{"{\"name\":\"Trillian3\",\"id\":\"5\",\"age\":\"33\"}"},
+            new Object[]{"{\"name\":\"Trillian4\",\"id\":\"5\",\"age\":\"33\"}"},
+            new Object[]{"{\"name\":\"Trillian5\",\"id\":\"5\",\"age\":\"33\"}"});
```
Comment (seut) [inline server/src/test/java/io/crate/execution/engine/collect/files/FileReadingIteratorTest.java:312]:
```suggestion
        List<Object[]> expectedResult = Arrays.asList(
            new Object[]{"{\"id\":\"1\"}"},
            new Object[]{"{\"id\":\"2\"}"},
            new Object[]{"{\"id\":\"3\"}"},
            new Object[]{"{\"id\":\"4\"}"},
            new Object[]{"{\"id\":\"5\"}"}
        );
```

File: server/src/test/java/io/crate/execution/engine/collect/files/FileReadingIteratorTest.java [hunk 7]
```diff
@@ -232,6 +234,85 @@ public String readLine() throws IOException {
         tester.verifyResultAndEdgeCaseBehaviour(expectedResult);
     }
 
+    @Test
```
Comment (seut) [inline server/src/test/java/io/crate/execution/engine/collect/files/FileReadingIteratorTest.java:237]: Not sure, but maybe worth to describe the reason here a bit more verbose.
```suggestion
    /**
     * Validates a bug which was resulting in duplicate reads of the same line when consecutive retries happen.
     */    
    @Test
```

File: docs/appendices/release-notes/unreleased.rst [hunk 8]
```diff
@@ -81,3 +81,7 @@ Fixes
 
 - Fixed an issue that caused the generated expressions on columns of type
   ``GEO_SHAPE`` not being evaluated on writes and such being ignored.
+
+- Fixed an issue that could generate duplicate data while executing
+  ``COPY FROM``. The trigger was a ``SocketTimeoutException`` being thrown
+  while recovering from another ``SocketTimeoutException``.
```
Comment (seut) [inline docs/appendices/release-notes/unreleased.rst:87]:
```suggestion
- Fixed an issue that could generate duplicate data on
  ``COPY FROM`` while some internal retries were happening
  due to I/O errors e.g. socket timeouts.
```

**Review by seut — COMMENTED** [review]
Good catch. Lets simplify the test scenario a bit.

File: server/src/test/java/io/crate/execution/engine/collect/files/FileReadingIteratorTest.java (hunk 7 above)
Comment (matriv) [reply server/src/test/java/io/crate/execution/engine/collect/files/FileReadingIteratorTest.java:237]: It's not bad to add more explanation, personally I'd prefer a link to the issue instead, but as you like.

**Review by matriv — APPROVED** [review]
LGTM, I agree with @seut's comments to simplify the test

File: server/src/test/java/io/crate/execution/engine/collect/files/FileReadingIteratorTest.java (hunk 7 above)
Comment (jeeminso) [reply server/src/test/java/io/crate/execution/engine/collect/files/FileReadingIteratorTest.java:237]: appending the link as well

**Comment by jeeminso** [comment]
Thank you :+1:

File: server/src/test/java/io/crate/execution/engine/collect/files/FileReadingIteratorTest.java (hunk 5 above)
Comment (jeeminso) [reply server/src/test/java/io/crate/execution/engine/collect/files/FileReadingIteratorTest.java]: missed this, will take care of it.

**Pull request merged**